sdks/go/pkg/beam/io/rtrackers/offsetrange/offsetrange_test.go

```diff
@@ -299,3 +301,299 @@ func TestTracker_TrySplit(t *testing.T) {
+	}
+}
+
+// TestGrowableTracker_SplitBeforeStart tests TrySplit() method for GrowableTracker
+// before claiming anything.
+func TestGrowableTracker_SplitBeforeStart(t *testing.T) {
+	estimator := offsetRangeEndEstimator{EstimateRangeEnd: 0}
+	rest := Restriction{Start: 0, End: math.MaxInt64}
+	tracker, err := NewGrowableTracker(rest, &estimator)
+	if err != nil {
+		t.Fatalf("error creating new GrowableTracker: %v", err)
+	}
+	estimator.SetEstimateRangeEnd(10)
+	p, r, _ := tracker.TrySplit(0)
+
+	expected := Restriction{0, 0}
+	if p.(Restriction) != expected {
+		t.Errorf("wrong primaries after TrySplit(0), got: %v, want: %v", p.(Restriction), expected)
+	}
+	if tracker.GetRestriction().(Restriction) != expected {
+		t.Errorf("wrong restriction tracked by tracker after TrySplit(0), got: %v, want: %v", tracker.GetRestriction().(Restriction), expected)
+	}
+	res := Restriction{0, math.MaxInt64}
+	if res != r.(Restriction) {
+		t.Errorf("wrong residual TrySplit(0), got: %v, want: %v", r.(Restriction), expected)
+	}
+}
+
+// TestGrowableTracker_CheckpointJustStarted tests TryClaim and TrySplit
+// for GrowableTracker.
+func TestGrowableTracker_CheckpointJustStarted(t *testing.T) {
+	estimator := offsetRangeEndEstimator{EstimateRangeEnd: 0}
+	rest := Restriction{Start: 0, End: math.MaxInt64}
+	tracker, err := NewGrowableTracker(rest, &estimator)
+	if err != nil {
+		t.Fatalf("error creating new GrowableTracker: %v", err)
+	}
+	if !tracker.TryClaim(int64(5)) {
+		t.Fatal("tracker.TryClaim(int64(5)) should've claimed.")
+	}
+	estimator.SetEstimateRangeEnd(0)
+	p, r, _ := tracker.TrySplit(0)
+	fmt.Printf("tracker: %#v", tracker)
+	if !tracker.IsDone() {
+		t.Fatal("tracker should have been done, but IsDone() returns false")
+	}
+
+	expPr := Restriction{0, 6}
+	if p.(Restriction) != expPr {
+		t.Errorf("wrong primaries after TrySplit(0), got: %v, want: %v", p.(Restriction), expPr)
+	}
+	if tracker.GetRestriction().(Restriction) != expPr {
+		t.Errorf("wrong restriction tracked by tracker after TrySplit(0), got: %v, want: %v", tracker.GetRestriction().(Restriction), expPr)
+	}
+	expRes := Restriction{6, math.MaxInt64}
+	if r.(Restriction) != expRes {
+		t.Errorf("wrong residual TrySplit(0), got: %v, want: %v", r.(Restriction), expRes)
+	}
+
+	tracker, err = NewGrowableTracker(rest, &estimator)
+	if err != nil {
+		t.Fatalf("error creating new GrowableTracker: %v", err)
+	}
+	if !tracker.TryClaim(int64(5)) {
+		t.Fatal("tracker.TryClaim(int64(5)) should've claimed.")
+	}
+	estimator.SetEstimateRangeEnd(20)
+	p, r, _ = tracker.TrySplit(0)
+	if !tracker.IsDone() {
+		t.Fatal("tracker should have been done, but IsDone() returns false")
+	}
+	if p.(Restriction) != expPr {
+		t.Errorf("wrong primaries after TrySplit(0), got: %v, want: %v", p.(Restriction), expPr)
+	}
+	if r.(Restriction) != expRes {
+		t.Errorf("wrong residual TrySplit(0), got: %v, want: %v", r.(Restriction), expRes)
+	}
+}
+
+func TestGrowableTracker_TrySplit(t *testing.T) {
```

Comment:
It feels like the various tests that call TrySplit with various initial conditions and then a mid condition adjustment should be able to become a Table Test. Instead of having a cases following each other, it should be identical to setting the initial conditions (eg the estimate, various bits of Claimed and not claimed state manually), and checking that the end results for each initial conditions are sound.

Yes, good testing practice recommends that we get there as much as possible via the public APIs (eg. TryClaim & TrySplit), but we don't need to be starting each from the 0 start, since we have well defined midpoints that can be cases themselves.  

This will simplify and consolidate the code for the tests.

Reply:
Deleting this test since Splits() is covering all of these